### PR TITLE
Add support for reading YAML from FILE *

### PR DESCRIPTION
**Pull request by sgallagher**
Fixes: https://github.com/fedora-modularity/libmodulemd/issues/50

Signed-off-by: Stephen Gallagher <sgallagh@redhat.com>

### CODE REVIEW

File: modulemd/modulemd-defaults.c [hunk 1]
```diff
@@ -524,6 +524,60 @@ modulemd_defaults_new_from_string (const gchar *yaml_string, GError **error)
...
+  GObject *object = NULL;
+  GPtrArray *data = NULL;
+  ModulemdDefaults *defaults = NULL;
+
+  if (!parse_yaml_stream (stream, &data, error))
+    {
+      return NULL;
+    }
+
+  for (gsize i = 0; i < data->len; i++)
```
Comment (ignatenkobrain) [inline modulemd/modulemd-defaults.c]: shouldn't it be `guint`? https://developer.gnome.org/glib/stable/glib-Pointer-Arrays.html#GPtrArray

File: modulemd/modulemd-defaults.c [hunk 2]
```diff
@@ -524,6 +524,60 @@ modulemd_defaults_new_from_string (const gchar *yaml_string, GError **error)
...
+    {
+      object = g_ptr_array_index (data, i);
+      if (MODULEMD_IS_DEFAULTS (object))
+        {
+          defaults = MODULEMD_DEFAULTS (g_object_ref (object));
```
Comment (ignatenkobrain) [inline modulemd/modulemd-defaults.c]: I don't think you need `MODULEMD_DEFAULTS` cast here... but not sure.

File: modulemd/modulemd-defaults.c [hunk 3]
```diff
@@ -524,6 +524,60 @@ modulemd_defaults_new_from_string (const gchar *yaml_string, GError **error)
...
+ * modulemd-defaults document in the given module stream. This must be freed
+ * with g_object_unref() when no longer needed.
+ *
+ * Since: 1.4
+ */
+ModulemdDefaults *
+modulemd_defaults_new_from_stream (FILE *stream, GError **error)
+{
+  GObject *object = NULL;
+  GPtrArray *data = NULL;
```
Comment (ignatenkobrain) [inline modulemd/modulemd-defaults.c]: `g_autoptr()`?

File: modulemd/modulemd-module.c [hunk 4]
```diff
@@ -3625,6 +3625,53 @@ modulemd_module_new_all_from_string_ext (const gchar *yaml_string,
...
+ * Return value: a new #ModulemdModule. When no longer needed, free it with
+ * g_object_unref().
+ *
+ * Since: 1.4
+ */
+ModulemdModule *
+modulemd_module_new_from_stream (FILE *stream)
+{
+  GError *error = NULL;
+  ModulemdModule *module = NULL;
```
Comment (ignatenkobrain) [inline modulemd/modulemd-module.c]: `g_autoptr`?

File: modulemd/modulemd-defaults.c (hunk 1 above)
Comment (sgallagher) [reply modulemd/modulemd-defaults.c]: It's irrelevant, really. gsize is essentially size_t which is pretty much designed for this purpose. It's also signed and always going to be at least as large as guint.

File: modulemd/modulemd-defaults.c (hunk 2 above)
Comment (sgallagher) [reply modulemd/modulemd-defaults.c]: I do, the compiler warns about it.

File: modulemd/modulemd-defaults.c (hunk 3 above)
Comment (sgallagher) [reply modulemd/modulemd-defaults.c]: Ah, I basically just copied this from the `modulemd_defaults_new_from_file()` routine. But you're right; I've been using `g_autoptr()` for all newly-written code; best not to break with that here. I will fix it up.

File: modulemd/modulemd-module.c (hunk 4 above)
Comment (sgallagher) [reply modulemd/modulemd-module.c]: Same as above.

**Comment by sgallagher** [comment]
I've updated this PR addressing the issues @ignatenkobrain saw, as well as the following:
* Fixed a couple error messages that were copied from the `_file()` versions to say "stream"
* Added a proper error return, even for functions whose file or string versions don't have it. That was a poor design that we're stuck with until 2.0, but there's no reason to keep it for new functions.

**Pull request closed without merging**